### ISSUE TITLE
Determination of SNAPSHOT should regex

### ISSUE THREAD

**Issue by briantopping** [sbt/sbt#2873]
https://github.com/sbt/sbt/search?utf8=✓&q=-SNAPSHOT gives a pretty concise picture that SBT determines a snapshot per the string literal `"-SNAPSHOT"`. This generally works, but an improvement would be to have it consider `"[-\.]SNAPSHOT"` in these cases.

The problem is OSGi resolvers do not see a version ending in "-SNAPSHOT" as valid. https://www.infoq.com/news/2012/04/osgi-snapshot is a detailed read of the subject for those interested.

Long story short, it would be wonderful if SBT could recognize OSGi's pedantry for snapshots and do the right thing. I haven't looked at the SBT code closely enough to know if this is feasible though. And the patch is so short that it's probably not worth the time to create it.

Could this be discussed?

Thanks, happy holidays. Brian

**Comment by eed3si9n** [comment]
I don't think anyone should consider a version ending in "-SNAPSHOT" as valid outside of a development machine or a CI environment. It causes all sorts of problems for sbt because people have different expectations about the semantics of SNAPSHOT, and it makes the dependency resolution performance horrendous if we try to default on the correctness. I don't want to add any feature that would encourage anyone to use SNAPSHOT beyond what it already does.

**Comment by briantopping** [comment]
Thanks Eugene. So are you against even accepting both "-SNAPSHOT" and ".SNAPSHOT"? 

I'm not expecting resolution outside of a dev or CI environment, I completely agree with what you are saying there.

Would you please reopen this for consideration by others? Thanks

**Comment by briantopping** [comment]
Another way to handle this is to just remove the `-` from the `-SNAPSHOT` strings.

**Comment by eed3si9n** [comment]
At least until sbt 1.0 comes out, I'm against adding feature to the behavior of dependency resolution semantics around `-SNAPSHOT`. @dwijnand can reopen it if he wants to weigh in, but my opinion is set on this. Also note this might require changing the behavior of sbt launcher 1.0.x and our Ivy fork.

**Comment by briantopping** [comment]
The justification is trying to see if SBT can be made a little easier to use in Karaf, which is a popular OSGi environment. Folks doing so will likely run into this bug, and right now, there's a widening gap between the tooling that *is* available in Maven for Karaf/OSGi support and the tooling that *is not* available for supporting Scala under Maven. 

That said, OSGi isn't mainstream and I wouldn't be surprised if this is rejected.

**Comment by eed3si9n** [comment]
You might want to lobby for the change on https://github.com/alexarchambault/coursier

**Comment by briantopping** [comment]
Ah, excellent, thanks for the pointer.

**Comment by alexarchambault** [comment]
https://github.com/coursier/coursier/commit/ab9341ac9bd12728a84667e9d72d88017e66442e should have fixed that.
